File: model-pipeline/voicetotext/acoustic/emissions.py
"""Emissions utilities: npz persistence and *views* over the (T, V) matrix.

Nothing here mutates or discards the matrix. ``greedy_decode`` is a convenience view; the
stored representation is always the full log-probability matrix.
"""

from __future__ import annotations

import unicodedata
from dataclasses import dataclass
from pathlib import Path

import numpy as np

from voicetotext.decoding.vocab import WORD_DELIMITER, blank_index, is_special
from voicetotext.types import Emissions, Hypothesis, WordSpan

# ----------------------------------------------------------------------------- persistence


def save_npz(em: Emissions, path: str | Path) -> Path:
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    np.savez_compressed(
        path,
        logprobs=em.logprobs.astype(np.float32),
        labels=np.array(em.labels, dtype=np.str_),
        frame_duration_ms=np.float64(em.frame_duration_ms),
        audio_id=np.str_(em.audio_id),
        model_id=np.str_(em.model_id),
    )
    return path


def load_npz(path: str | Path) -> Emissions:
    with np.load(Path(path), allow_pickle=False) as z:
        return Emissions(
            logprobs=np.ascontiguousarray(z["logprobs"], dtype=np.float32),
            labels=[str(x) for x in z["labels"]],
            frame_duration_ms=float(z["frame_duration_ms"]),
            audio_id=str(z["audio_id"]),
            model_id=str(z["model_id"]),
        )


# ----------------------------------------------------------------------------- greedy view


@dataclass
class TokenSpan:
    """One collapsed CTC token on the argmax path."""

    label_id: int
    label: str
    start_frame: int
    end_frame: int  # exclusive
    prob: float  # mean posterior of the token over its frames


def greedy_path(em: Emissions) -> list[TokenSpan]:
    """Argmax per frame, collapse repeats, drop blanks. Keeps the word delimiter as a token."""
    blank = blank_index(em.labels)
    ids = em.logprobs.argmax(axis=1)
    maxp = np.exp(em.logprobs.max(axis=1))
    spans: list[TokenSpan] = []
    t = 0
    T = len(ids)
    while t < T:
        lab = int(ids[t])
        start = t
        while t < T and int(ids[t]) == lab:
            t += 1
        if lab != blank:
            spans.append(
                TokenSpan(
                    label_id=lab,
                    label=em.labels[lab],
                    start_frame=start,
                    end_frame=t,
                    prob=float(maxp[start:t].mean()),
                )
            )
    return spans


def _is_text_token(label: str) -> bool:
    return not is_special(label)


def greedy_words(em: Emissions) -> list[WordSpan]:
    """Words on the greedy path with frame spans and a simple confidence: the mean over the
    word's characters of each character's mean posterior."""
    words: list[WordSpan] = []
    chars: list[TokenSpan] = []

    def flush() -> None:
        if chars:
            words.append(
                WordSpan(
                    text=unicodedata.normalize("NFC", "".join(c.label for c in chars)),
                    start_frame=chars[0].start_frame,
                    end_frame=chars[-1].end_frame,
                    confidence=float(np.mean([c.prob for c in chars])),
                )
            )
            chars.clear()

    for tok in greedy_path(em):
        if tok.label == WORD_DELIMITER or tok.label == " ":
            flush()
        elif _is_text_token(tok.label):
            chars.append(tok)
    flush()
    return words


def greedy_decode(em: Emissions) -> str:
    """Unconstrained acoustic string. A view; never store it in place of the emissions."""
    return " ".join(w.text for w in greedy_words(em))


def greedy_hypothesis(em: Emissions) -> Hypothesis:
    words = greedy_words(em)
    return Hypothesis(
        text=" ".join(w.text for w in words),
        acoustic_score=float(em.logprobs.max(axis=1).sum()),
        lm_score=0.0,
        words=words,
    )


def posteriors(em: Emissions) -> np.ndarray:
    """(T, V) probabilities. Allocates; use sparingly."""
    return np.exp(em.logprobs)

File: android-app/app/src/main/java/com/openhackathon/voicetotext/asr/Emissions.kt
package com.openhackathon.voicetotext.asr

import kotlin.math.exp
import kotlin.math.ln

/**
 * The (T, V) log-probability matrix produced by the CTC acoustic layer.
 *
 * This is the unit of work. Argmax strings are *views*
 * computed on demand by [greedyDecode]; they are never the stored representation, because
 * everything above the acoustic layer (beam search, speaker adaptation, correction) needs
 * the full distribution, not the winning character.
 *
 * @param logProbs row-major, length T * V.
 * @param labels index-aligned to V.
 */
class Emissions(
    val logProbs: FloatArray,
    val labels: List<String>,
    val frameDurationMs: Float = 20f,
    val audioId: String = "",
    val modelId: String = "",
) {
    val vocabSize: Int = labels.size
    val numFrames: Int = logProbs.size / vocabSize

    init {
        require(logProbs.size % vocabSize == 0) {
            "logProbs length ${logProbs.size} is not a multiple of V=$vocabSize"
        }
    }

    val durationSeconds: Float get() = numFrames * frameDurationMs / 1000f

    fun logProb(frame: Int, label: Int): Float = logProbs[frame * vocabSize + label]

    fun argmaxAt(frame: Int): Int {
        val base = frame * vocabSize
        var best = 0
        var bestVal = logProbs[base]
        for (v in 1 until vocabSize) {
            val x = logProbs[base + v]
            if (x > bestVal) { bestVal = x; best = v }
        }
        return best
    }

    fun maxProbAt(frame: Int): Float = exp(logProbs[frame * vocabSize + argmaxAt(frame)])

    val blankIndex: Int by lazy {
        labels.indexOfFirst { it == "[PAD]" || it == "<pad>" }.also {
            require(it >= 0) { "no blank token in labels" }
        }
    }

    private fun isSpecial(label: String): Boolean =
        label == "[PAD]" || label == "<pad>" || label == "[UNK]" || label == "<unk>" ||
            label == "<s>" || label == "</s>" || label.startsWith("<unused")

    /** One collapsed CTC token on the argmax path. */
    data class TokenSpan(val label: String, val startFrame: Int, val endFrame: Int, val prob: Float)

    /**
     * Argmax per frame, then the CTC collapse rule: merge runs of the same label, drop blanks.
     * This is a deterministic decoding rule, not a correction: it has no vocabulary and no
     * language model, so it will happily emit non-words.
     */
    fun greedyPath(): List<TokenSpan> {
        val out = ArrayList<TokenSpan>()
        var t = 0
        while (t < numFrames) {
            val id = argmaxAt(t)
            val start = t
            var sum = 0f
            while (t < numFrames && argmaxAt(t) == id) { sum += maxProbAt(t); t++ }
            if (id != blankIndex) out.add(TokenSpan(labels[id], start, t, sum / (t - start)))
        }
        return out
    }

    data class WordSpan(val text: String, val startFrame: Int, val endFrame: Int, val confidence: Float)

    fun greedyWords(): List<WordSpan> {
        val words = ArrayList<WordSpan>()
        val chars = ArrayList<TokenSpan>()
        fun flush() {
            if (chars.isNotEmpty()) {
                words.add(
                    WordSpan(
                        chars.joinToString("") { it.label },
                        chars.first().startFrame,
                        chars.last().endFrame,
                        chars.map { it.prob }.average().toFloat(),
                    )
                )
                chars.clear()
            }
        }
        for (tok in greedyPath()) {
            if (tok.label == "|" || tok.label == " ") flush() else if (!isSpecial(tok.label)) chars.add(tok)
        }
        flush()
        return words
    }

    fun greedyDecode(): String = greedyWords().joinToString(" ") { it.text }

    /** Fraction of frames whose argmax is the blank token. High values mean sparse speech. */
    fun blankFraction(): Float {
        var n = 0
        for (t in 0 until numFrames) if (argmaxAt(t) == blankIndex) n++
        return n.toFloat() / numFrames
    }

    companion object {
        /** Row-wise log-softmax of raw logits, in place. */
        fun fromLogits(
            logits: FloatArray,
            labels: List<String>,
            frameDurationMs: Float = 20f,
            audioId: String = "",
            modelId: String = "",
        ): Emissions {
            val v = labels.size
            val frames = logits.size / v
            for (t in 0 until frames) {
                val base = t * v
                var max = logits[base]
                for (i in 1 until v) if (logits[base + i] > max) max = logits[base + i]
                var sum = 0f
                for (i in 0 until v) sum += exp(logits[base + i] - max)
                val logZ = max + ln(sum)
                for (i in 0 until v) logits[base + i] -= logZ
            }
            return Emissions(logits, labels, frameDurationMs, audioId, modelId)
        }
    }
}
